File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack=[]
        maxA=0
        n=len(heights)
        for i in range(n+1):
            while stack and (i==n or heights[stack[-1]]>=heights[i]):
                height=heights[stack[-1]]
                stack.pop()
                width=0
                if not stack: width = i
                else: width=i-stack[-1]-1
                maxA=max(maxA,width*height)
            stack.append(i)
        return maxA

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
  public:
    int largestRectangleArea(vector < int > & histo) {
      stack < int > st;
      int maxA = 0;
      int n = histo.size();
      for (int i = 0; i <= n; i++) {
        while (!st.empty() && (i == n || histo[st.top()] >= histo[i])) {
          // we are processing the already passed buildings
          int height = histo[st.top()];
          st.pop();
          int width;
          if (st.empty())
            width = i;
          else
            width = i - st.top() - 1;
          maxA = max(maxA, width * height);
            // cout<<i<<" "<<height<<" "<<width<<endl;
        }
        st.push(i);
      }
      return maxA;
    }
};